Source/WebGPU/WGSL/Types.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Types.h"
+
+namespace WGSL {
+
+void printInternal(PrintStream& out, const Type& type)
+{
+    WTF::switchOn(type,
+        [&](const Variable& variable) {
+            out.print("type_variable", variable.id);
+        },
+        [&](const Primitive& primitive) {
+            switch (primitive.kind) {
+#define PRIMITIVE_CASE(kind, name) \
```

Comment:
Is this really better than just listing them all out?

If we update the list in the future and forget to update this place, we'll get a compilation error...

Reply:
Are you referring to the macro or the `switchOn`?

For the macro, I personally prefer it over having an enum and then having the switch here, but I don't feel strongly about it, just think it's easier to keep the definition + string representation in the same place.

The reason I asked about the `switchOn` is because I don't see how updating the macro would break this.